packages/block-library/src/block/edit.js

```diff
@@ -67,23 +87,109 @@ const useInferredLayout = ( blocks, parentLayout ) => {
 	}, [ blocks, parentLayout ] );
 };
 
+function applyInitialDynamicContent(
+	blocks,
+	dynamicContent = {},
+	defaultValues
+) {
+	return blocks.map( ( block ) => {
+		const innerBlocks = applyInitialDynamicContent(
+			block.innerBlocks,
+			dynamicContent,
+			defaultValues
+		);
+		if ( ! isPartiallySynced( block ) ) return { ...block, innerBlocks };
+		const attributes = getPartiallySyncedAttributes( block );
+		const newAttributes = { ...block.attributes };
+		for ( const [ attributeKey, id ] of Object.entries( attributes ) ) {
+			defaultValues[ id ] = block.attributes[ attributeKey ];
+			if ( dynamicContent[ id ] ) {
+				newAttributes[ attributeKey ] = dynamicContent[ id ];
+			}
+		}
+		return {
+			...block,
+			attributes: newAttributes,
+			innerBlocks,
+		};
+	} );
+}
+
+function getDynamicContentFromBlocks( blocks, defaultValues ) {
+	/** @type {Record<string, unknown>} */
+	const dynamicContent = {};
+	for ( const block of blocks ) {
+		Object.assign(
+			dynamicContent,
+			getDynamicContentFromBlocks( block.innerBlocks, defaultValues )
+		);
+		if ( ! isPartiallySynced( block ) ) continue;
+		const attributes = getPartiallySyncedAttributes( block );
+		for ( const [ attributeKey, id ] of Object.entries( attributes ) ) {
+			if ( block.attributes[ attributeKey ] !== defaultValues[ id ] ) {
+				dynamicContent[ id ] = block.attributes[ attributeKey ];
+			}
+		}
+	}
+	return Object.keys( dynamicContent ).length > 0
+		? dynamicContent
+		: undefined;
+}
+
 export default function ReusableBlockEdit( {
 	name,
-	attributes: { ref },
+	attributes: { ref, dynamicContent },
 	__unstableParentLayout: parentLayout,
+	clientId: patternClientId,
 } ) {
+	const registry = useRegistry();
 	const hasAlreadyRendered = useHasRecursion( ref );
 	const { record, hasResolved } = useEntityRecord(
 		'postType',
 		'wp_block',
 		ref
 	);
 	const isMissing = hasResolved && ! record;
+	const initialDynamicContent = useRef( dynamicContent );
+	const defaultValuesRef = useRef( {} );
+	const {
+		replaceInnerBlocks,
+		__unstableMarkNextChangeAsNotPersistent,
+		setBlockEditingMode,
+	} = useDispatch( blockEditorStore );
+	const { getBlockEditingMode } = useSelect( blockEditorStore );
 
-	const [ blocks, onInput, onChange ] = useEntityBlockEditor(
-		'postType',
-		'wp_block',
-		{ id: ref }
+	useEffect( () => {
+		if ( ! record?.content?.raw ) return;
+		const initialBlocks = parse( record.content.raw );
+
+		const editingMode = getBlockEditingMode( patternClientId );
+		registry.batch( () => {
+			setBlockEditingMode( patternClientId, 'default' );
+			__unstableMarkNextChangeAsNotPersistent();
+			replaceInnerBlocks(
+				patternClientId,
+				applyInitialDynamicContent(
+					initialBlocks,
+					initialDynamicContent.current,
+					defaultValuesRef.current
+				)
+			);
```

Comment:
I guess it relies on the pattern not being editable from the editor?

I can't see code that ever saves changes to the inner blocks back to the entity, so that's my assumption. I guess it means there's a risk that if #56534 isn't successful, this approach might need to be adjusted.

Reply:
Yeah, it's under the assumption that we'll need a separate UI to edit the pattern source. This approach also breaks the current behavior, but I think that's expected from the product's vision. We'll need more UI/UX work to make it acceptable for regular users though 😅.